### ISSUE TITLE
FN: iter_cloned_collect with a large array

### ISSUE THREAD

**Issue by Y-Nak** [rust-lang/rust-clippy#6808]
Lint name: `iter_cloned_collect`

I tried this code:

```rust
let arr: [u8; 64] = [0; 64];
let v: Vec<_> = arr.iter().cloned().collect();
```

I expected to see this happened: `iter_cloned_collect` is triggered.

Instead, this happened:  nothing.

### Meta

- `cargo clippy -V`: clippy 0.0.212 (cb75ad5 2021-02-10)
- `rustc -Vv`: 
```
rustc 1.50.0 (cb75ad5db 2021-02-10)
binary: rustc
commit-hash: cb75ad5db02783e8b0222fee363c5f63f7e2cf5b
commit-date: 2021-02-10
host: x86_64-unknown-linux-gnu
release: 1.50.0
 ```

--- 

I think this is caused by [`derefs_to_slice`](https://github.com/rust-lang/rust-clippy/blob/7154b2360114977cd67bb844114cf429fa9cdb88/clippy_lints/src/methods/mod.rs#L2775-L2813) and I suppose we should use `expr_ty_adjusted` instead of [`may_slice`](https://github.com/rust-lang/rust-clippy/blob/7154b2360114977cd67bb844114cf429fa9cdb88/clippy_lints/src/methods/mod.rs#L2780-L2791) to handle `coercion` in the function.
